README.md
<img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Life cycle: experimental"> <a href="https://godoc.org/github.com/openanno/bget"><img src="https://godoc.org/github.com/openanno/bget?status.svg" alt="GoDoc"></a>

# bget

bget is an portable tool with several sub-commands to query bioinformatics APIs, data, databases and files. The Golang `http` library, `wget`, `curl`, `axel`, `git`, and `rsync` were supported as the query engine.

Supported types:

- Reference genomes
- Source code of bioinformatics tools
- Bioinformatics databases and files
- Papers material
- ......

Downstream tool:

- [bioctl](https://github.com/openanno/bioctl): convert, format, and other functions
- [bioextr](https://github.com/openanno/bioextr): text-mining functions

## Prerequisities

For website spider (optional):

- Headless Chrome is required for some of website with JavaScript driven render pages. For windows users, you may need to create an alias of Chrome to make [chromedp](https://github.com/chromedp/chromedp) work.

For raw sequencing data query (optional):

- [sra-tools](https://github.com/ncbi/sra-tools) for SRA and dbGAP database: `bget i sratools`;
- [pyega3](https://github.com/EGA-archive/ega-download-client) for EGA database: `pip3 install pyega3`;
- [gdc-client](https://gdc.cancer.gov/access-data/gdc-data-transfer-tool) for GDC portal: `bget i gdc-client@1.5.0 -u`.

## Installation

```bash
# windows
wget https://github.com/openanno/bget/releases/download/v0.3.0/bget.exe

# osx
wget https://github.com/openanno/bget/releases/download/v0.3.0/bget_osx
mv bget_osx bget
chmod a+x bget

# linux
wget https://github.com/openanno/bget/releases/download/v0.3.0/bget_linux64
mv bget_linux64 bget
chmod a+x bget

# get latest version
go get -u github.com/openanno/bget
```

## Usage

Command line outputs see [here](https://openbiox.github.io/bget/cli.html)

### Query webiste API

`bget api` can be used to query serveral website APIs, such as PubMed, Datasetdataset2tools, and GDC portal website.

In addition, you can use the downstream tool [bioctl](https://github.com/openanno/bioctl) to conduct the simple text-mining of PubMed abstract at the sentence level.

```bash
# NCBI eutils
# query pubmed
bget api ncbi -d pubmed -q B-ALL --format XML -e your_email@domain.com

# query pubmed and convert it to json format that also extract all URLs and calculate the words connections
bget api ncbi -q "Galectins control MTOR and AMPK in response to lysosomal damage to induce autophagy OR MTOR-independent autophagy induced by interrupted endoplasmic reticulum-mitochondrial Ca2+ communication: a dead end in cancer cells. OR The PARK10 gene USP24 is a negative regulator of autophagy and ULK1 protein stability OR Coordinate regulation of autophagy and the ubiquitin proteasome system by MTOR." | bioctl cvrt --xml2json pubmed -

# datasetdataset2tools API
# query canned analysis accession	, e.g. DCA00000060.
bget api dta -a DCA00000060
# query dataset accession number, e.g. GSE31106 
bget api dta -s GSE31106 | bioctl fmt --json-pretty -
# query via object type
bget api dta --type dataset | bioctl fmt --json-pretty --indent 2 -
# props of dataset accession, e.g. upregulated.
bget api dta -g upregulated | json2csv -o out.csv

# GDC portal API
# retrive projects meta info from GDC portal
bget api gdc -p
bget api gdc -p --json-pretty
bget api gdc -p -q TARGET-NBL --json-pretty
bget api gdc -p --format tsv > tcga_projects.tsv
bget api gdc -p --format csv > tcga_projects.csv
bget api gdc -p --from 1 --szie 2
# check GDC portal status (https://portal.gdc.cancer.gov/)
bget api gdc -s
# retrive cases info from GDC portal
bget api gdc -c
# retrive files info from GDC portal
bget api gdc -f
# retrive annotations info from GDC portal
bget api gdc -a
# query manifest for gdc-client
bget api gdc -m -q "5b2974ad-f932-499b-90a3-93577a9f0573,556e5e3f-0ab9-4b6c-aa62-c42f6a6cf20c" -o my_manifest.txt
bget api gdc -m -q "5b2974ad-f932-499b-90a3-93577a9f0573,556e5e3f-0ab9-4b6c-aa62-c42f6a6cf20c" > my_manifest.txt
bget api gdc -m -q "5b2974ad-f932-499b-90a3-93577a9f0573,556e5e3f-0ab9-4b6c-aa62-c42f6a6cf20c" -n
# query data that only support the samll filesize
bget api gdc -d -q "5b2974ad-f932-499b-90a3-93577a9f0573" -n

# clinicaltrials.gov API
# returns the date when the ClinicalTrials.gov dataset was posted.
bget api cligov --info-dat-vers
# returns the current version number of the ClinicalTrials.gov API
bget api cligov --info-api-vers
# returns detailed definitions.
bget api cligov --info-api-defs
# returns all available data elements for a single study record.
bget api cligov --info-study-struct
# returns all data elements.
bget api cligov --info-study-fields
# returns an annotated version of the Study Structure info URL.
bget api cligov --info-study-stat
# returns groups of weighted study fields, or "search areas"
bget api cligov --info-search-area
	
# query functions
bget api cligov -q heart+attack --full-studies --format json
bget api cligov -q heart+attack --fields NCTId,Condition,BriefTitle --study-fields
bget api cligov -q heart+attack --field Condition --field-values

# bio.tools API
# query item detail
bget api biots --tool signalp

# search item
bget api biots --name signalp
bget api biots --topic Proteomics
bget api biots --dtype 'Protein sequence'
bget api biots --dfmt FASTA
bget api biots --ofmt 'ClustalW format'

# crossref
bget api crf --doi 10.1073/pnas.1814397115
bget api crf --doi 10.1073/pnas.1814397115 --xml2json --json-pretty --indent 1

# mgrast
bget api mgrast anno --info
# retrieval of SwissProt taxonomy annotations with a cut-off 10^100 for dataset mgm4447943.3
bget api mgrast anno --evalue 100 --type organism --source SwissProt --seq mgm4447943.3

bget api mgrast compute --info

# returns all data in the system. Warning: this request returns 8MB+ and takes 5+ seconds
bget api covid19 --all
# returns all countries and associated provinces
bget api covid19 --cts
# returns all cases by case type for a country from the first recorded case.
bget api covid19 --ct --name China
# returns all cases by case type for a country from the first recorded case.
bget api covid19 --ct-d-one --name China
# returns all cases by case type for a country.
bget api covid19 --ct-d-one-total --name China
# returns all cases by case type for a country from the first recorded case with the latest record being the live count.
bget api covid19 --ct-st-d-one --name China --status confirmed
# returns all cases by case type for a country from the first recorded case.
bget api covid19 --ct-st-d-one-live --name China --status confirmed
# returns all cases by case type for a country from the first recorded case
bget api covid19 --ct-st-d-one-total --name China --status confirmed
# returns all cases by case type for a country with the latest record being the live count.
bget api covid19 --ct-st-live --name China --status confirmed
# returns all cases by case type for a country.
bget api covid19 --ct-st --name China --status confirmed
# returns all cases by case type for a country.
bget api covid19 --ct-st-total --name China --status confirmed
# returns all cases of a country.
bget api covid19 --ct-total --name China
# returns all live cases by case type for a country.
bget api covid19 --live-ct --name China
# returns all live cases by case type for a country after a given date.
bget api covid19 --live-ct-st-date --name China --status confirmed --date 2020-04-20T06:20:47Z
# returns all live cases by case type for a country.
bget api covid19 --live-ct-st --name China --status confirmed
# a summary of new and total cases per country
bget api covid19 --summary
bget api covid19 --export
bget api covid19 --webhook https://your_webhook.com
```

### Query DOI resources

`bget doi` can be used to query DOI resources from website and journals that the supported items are continuely increasing.

```bash
## query zendo website with 3 thread
bget doi 10.5281/zenodo.3363060 10.5281/zenodo.3357455 10.5281/zenodo.3351812 -t 3

## query fulltext of publications (proxy may needed)
bget doi 10.1016/j.devcel.2017.03.001 10.1016/j.stem.2019.07.009 10.1016/j.celrep.2018.03.072 -t 2

## query publications with supplementary files
bget doi 10.1038/s41586-019-1844-5 --suppl

# query pdf and meta data using PubMed ID
dois=`bget api ncbi --xml2json --json-pretty -q '30487223[pmid] or 30402350[pmid] or 29279377[pmid]' --size 3 -m 3 | grep / | grep 10. | sed 's/ .* "//' | tr -d '",' | sort -u` && echo ${dois} && bget doi ${dois} --print-meta --print-crossref
```

We can query PDF of the manuscript via using Endnote or sci-hub. However, you can not easily get the supplementary files of scientific papers based on the two ways.

![doi demo](https://github.com/openanno/bget/raw/master/docs/static/doi.gif)

Here, we are developing and sharing an open-source tool bget with `doi` subcommand to query supplementary files of scientific papers. The journals with high impact factors or those integrative publishers are a higher priority in our development plan, see [here](http://openbiox.github.io/bget/doi.html)

**Warn**: It is noted that we do not want to distribute any pirated resources or cause unnecessary network congestion. We hope this tool can provide an optional method to more easily query related files of scientific papers. Please use it in a non-invasive way (i.e. high concurrency, long continuous request). If you do not follow the policies of the relevant website (i.e. continuous download or limited copyright), you will lose the authorization to use this tool.

### Query files via alias key

`bget i` can be used to query a set of files via the alias key, such as bwa, samtools, reffa/defuse, and db/annovar.

```bash
# download bwa source (with task env info)
bget i bwa --verbose 2
# get all available keys
bget i -a
# in JSON format
bget i -a --format json
# view all bwa and samtools available tags in table
bget i bwa samtools -v
# view all bwa and samtools available tags in json
bget i bwa samtools -v --format json

# force download defuse reference (with task env info and save log to file)
bget i "reffa/defuse@GRCh38 #97" -t 10 -f
bget i reffa/defuse@GRCh38 release=97 -t 10 -f
# download annovar reference
bget i db/annovar@clinvar_20170501 db/annovar@clinvar_20180603 builder=hg38

bget i db/annovar -v --formt text
bget i db/annovar version='clinvar_20131105, clinvar_20140211, clinvar_20140303, clinvar_20140702, clinvar_20140902, clinvar_20140929, clinvar_20150330, clinvar_20150629, clinvar_20151201, clinvar_20160302, clinvar_20161128, clinvar_20170130, clinvar_20170501, clinvar_20170905, clinvar_20180603, avsnp150, avsnp147, avsnp144, avsnp142, avsnp138, cadd, caddgt10, caddgt20, cadd13, cadd13gt10, cadd13gt20, cg69, cg46, cosmic70, cosmic68wgs, cosmic68, cosmic67wgs, cosmic67, cosmic65, cosmic64, dbnsfp35a, dbnsfp33a, dbnsfp31a_interpro, dbnsfp30a, dbscsnv11, eigen, esp6500siv2_ea, esp6500siv2_aa, esp6500siv2_all, exac03nontcga, exac03nonpsych, exac03, fathmm, gerp++gt2, gme, gnomad_exome, gnomad_genome, gwava, hrcr1, icgc21, intervar_20170202, kaviar_20150923, ljb26_all, mcap, mitimpact2, mitimpact24, nci60, popfreq_max_20150413, popfreq_all_20150413, revel, regsnpintron' builder=hg19 -t 10 -f
```

### Query FASTQ/CEL files from GEO/SRA/EGA/dbGAP/GDC

`bget seq` can be used to query files from [Gene Expression Omnibus (GEO)](https://www.ncbi.nlm.nih.gov/geo), [Sequence Read Archive (SRA)](https://www.ncbi.nlm.nih.gov/sra/), and [GDC Data Portal](https://portal.gdc.cancer.gov/).

```bash
# download files from SRA databaes using prefetch
bget seq ERR3324530 SRR544879

# download files from GEO databaes, auto download SRA acc list and run info
bget seq GSE23543 GSM1098572 -t 2

# download files from dbGap database using krt files
bget seq dbgap.krt using prefetch

# download dataset from EGA databaes using pyega3
bget seq EGAD00001000951

# download file from EGA databaes using pyega3
bget seq EGAF00000585895

# download TCGA files using file id using gdc-client
bget seq b7670817-9d6b-494e-9e22-8494e2fd430d

# download TCGA files using manifest files using gdc-client
# split for parallel
split -a 3 --additional-suffix=.txt -l 100 gdc_manifest.2019-08-23-TCGA.txt -d
for i in x*.txt
do
  head -n 1 x000.txt > ${i}.tmp && cat ${i} >> ${i}.tmp &&mv ${i}.tmp ${i}
done
sed -i '1d' x000.txt
bget seq *.txt -t 5

# support auto (if you do not have *.krt, TCGA manifest, please not include it for test)
bget seq SRR544879 GSE23543 EGAD00001000951 b7670817-9d6b-494e-9e22-8494e2fd430d dbgap.krt *.txt -t 5
```

### Query URLs

`bget url` can be used to query files using URLs.

```bash
urls="https://dldir1.qq.com/weixin/Windows/WeChatSetup.exe,http://download.oray.com/pgy/windows/PgyVPN_4.1.0.21693.exe,https://dldir1.qq.com/qqfile/qq/PCQQ9.1.6/25786/QQ9.1.6.25786.exe" && echo $urls | tr "," "\n"> /tmp/urls.list

bget url ${urls}
bget url https://dldir1.qq.com/weixin/Windows/WeChatSetup.exe https://dldir1.qq.com/qqfile/qq/PCQQ9.1.6/25786/QQ9.1.6.25786.exe --save-log
bget url ${urls} -t 3 -o /tmp/download -f -g wget --save-log --verbose 2
bget url ${urls} -t 2 -o /tmp/download --save-log --verbose 2
bget url ${urls} -t 3 -o /tmp/download -g wget --resume
bget url -l /tmp/urls.list -o /tmp/download -f -t 3

# query github repo (support assets files)
bget url Miachol/github_demo --github
bget url PapenfussLab/gridss openanno/bget --with-github-assets -t 5 --github
bget url PapenfussLab/gridss openanno/bget --only-github-assets -t 5 --github
bget url PapenfussLab/gridss openanno/bget --with-github-assets --with-assets-versions v2.7.2,v0.1.3 -t 5 --github
```

## Maintainer

- [@Jianfeng](https://github.com/Miachol)

## License

Academic Free License version 3.0



ChangeLog
2020-06-05 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* auto load http_proxy or https_proxy
		* fix defuse genomeVersionConvertor

2020-04-29 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support covid19 api

2020-04-22 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* bio.tools API support --from and --size flag

2020-04-21 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support URL with multi-thread request
		* default support resume
		* --ignore => --resume

2020-04-20 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support custom links
		* rename key subcmd to i

2020-03-31 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* new '--print-meta' flag in `bget doi`
		* new '--print-crossref' flag in `bget doi`
		* new '--xml2json' flag in `bget api`
		* add `crf` subcommand in `bget api`
		* expand suported doi list

2020-03-28 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* fix peerj doi query
		* add initColley function in spider

2020-03-27 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support 10.35946, 10.20882, 10.17712, 10.17645, 10.14309, 10.7326,
		10.7189, 10.7185, 10.5853, 10.5694, 10.5670, 10.5598, 10.4158,
		10.4155, 10.4110, 10.3982, 10.3835, 10.3386, 10.3348, 10.3238,
		10.3233, 10.3168, 10.2807, 10.2471, 10.2340, 10.2136, 10.2113,
		10.2105, 10.1681, 10.1556, 10.1513, 10.1373, 10.1289, 10.1287, 
		10.1257, 10.1246, 10.1210, 10.1176, 10.1137, 10.1130, 10.1107,
		10.1096, 10.1094, 10.1042
		* add functions (AddPdfSpider, AddPdfplusSpider, DirectSpider,
		AddPdfplusWithSupplSpider, AddPdfWithSupplSpider, AddLastPdfSpider,
		AddDownloadSpider, KoreaMedSpider, ReplaceHtmlSpider, and
		RetriveRedirectLink) in spider/universal.go
		* mgrast api add library, proj, and sample command

2020-03-26 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support 10.3233 and 10.7185
		* add r in `bget key`

2020-03-25 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support bget api mgrast (anno, compute, darkmatter, download modules)
		* support 10.1152, 10.1098, 10.1175, 10.21037, 10.2196, 10.1055,
		10.1086, 10.1145, 10.2217, 10.7150

2020-03-24 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support journals.lww.com (10.1097)
		* support journals.sagepub.com (10.1177)
		* support www.liebertpub.com (10.1089)
		* support ard.bmj.com, adc.bmj.com, casereports.bmj.com, ebm.bmj.com,
		gh.bmj.com, informatics.bmj.com, innovations.bmj.com,
		bmjleader.bmj.com, militaryhealth.bmj.com, neurologyopen.bmj.com,
		nutrition.bmj.com, bmjopen.bmj.com, drc.bmj.com,
		bmjopengastro.bmj.com, bmjophth.bmj.com, qir.bmj.com,
		bmjopenrespres.bmj.com, openscience.bmj.com, bmjopensem.bmj.com,
		qualitysafety.bmj.com, bmjpaedsopen.bmj.com, srh.bmj.com,
		stel.bmj.com, spcare.bmj.com, sit.bmj.com, bjo.bmj.com, bjsm.bmj.com,
		considerations.bmj.com, dtb.bmj.com, ep.bmj.com, emj.bmj.com,
		esmoopen.bmj.com, ejhp.bmj.com, ebmh.bmj.com, ebn.bmj.com,
		fmch.bmj.com, fn.bmj.com, fg.bmj.com, gpsych.bmj.com, heart.bmj.com,
		heartasia.bmj.com, injuryprevention.bmj.com, inpractice.bmj.com,
		ihj.bmj.com, ijgc.bmj.com, jitc.bmj.com, jcp.bmj.com, jech.bmj.com,
		jim.bmj.com, jisakos.bmj.com, jme.bmj.com, jmg.bmj.com, jnnp.bmj.com,
		jnis.bmj.com, lupus.bmj.com, mh.bmj.com, oem.bmj.com,
		openheart.bmj.com, pmj.bmj.com, pn.bmj.com, rapm.bmj.com,
		rmdopen.bmj.com, sti.bmj.com, svn.bmj.com, thorax.bmj.com,
		tobaccocontrol.bmj.com, tsaco.bmj.com, veterinaryrecord.bmj.com,
		vetrecordcasereports.bmj.com, vetrecordopen.bmj.com, wjps.bmj.com
		(10.1136)

2020-03-23 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support cellimagelibrary.org (10.7295)
		* support ieeexplore.ieee.org (10.1109)
		* add GWAS Catalog db links: `bget key db/gwas-catalog -n`
		* rename --pmc flag to --enable-pmc in `bget doi`
		* add --enable-scihub flag in `bget doi`

2020-01-18 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* --task-id add -k
		* move convert and text-mining function to bioctl and bioextr pkg
		* support git version parse in `bget key` tools

2020-01-16 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* improve the log output
		* support 'bget key -a' output to json format

2020-01-11 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support clinicaltrials.gov API
		* support bio.tools API
		* add --extra flag in api
		* add db/clingov: ClinicalTrials.gov/AllAPIJSON.zip

2020-01-10 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* hide global flags in the rootCmd
		* integrate github subcmd in url subcmd
		* integrate api subcmd

2019-12-03 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* add "--only-assets" in github subcmd

2019-12-01 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* add pubmed github links in key subcmd
		* parse key subcmd with case-insensitive and "_" => "-"
		* add bitbucket version API
		* support github assets in key subcmd
		* add github subcmd

2019-11-29 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* add more db items in key subcmd
		* set default version var

2019-11-29 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* classify items in key subcmd

2019-11-28 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* add wdl workflows in 'bget key' mode
		* add openbiox/awosome-bioinformatics github links

2019-11-19 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support www.sciencedirect.com supplementary files

2019-11-09 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support pubs.acs.org (test: 10.1021/acs.chemrev.9b00372)
		* support jamanetwork.com (test: 10.1001/jama.2019.17379)
		* support pubs.rsc.org (test: 10.1039/C9CS00377K)
		* support journals.aps.org (test: 10.1103/RevModPhys.91.045001)
		* support www.annualreviews.org (test: 10.1146/annurev-astro-091918-104446)

2019-11-08 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* support figshare.com (test: 10.6084/m9.figshare.6634136)

2019-11-07 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>

		* update annovar pool
		* support DIABETES, DEVELOPMENT, and Microbial Genomics via UniversalSpider
		* support OncoImmunology (update TandfonlineSpider)
		* support Stem Cells via WileyComSpider

2019-11-06 Li Jianfeng <lee_jianfeng@sjtu.edu.cn>


cmd/doi.go
package cmd

import (
	//"bufio"
	//"bytes"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	neturl "net/url"
	"path"
	"strings"
	"time"

	"github.com/openanno/bget/api/fetch"
	"github.com/openanno/bget/api/types"
	"github.com/openanno/bget/spider"
	cio "github.com/openbiox/ligo/io"
	cnet "github.com/openbiox/ligo/net"
	"github.com/openbiox/ligo/slice"
	stringo "github.com/openbiox/ligo/stringo"
	"github.com/spf13/cobra"
	"github.com/tidwall/pretty"
)

var fullText string
var suppl bool
var pmc bool
var universeSpider bool
var scihub bool
var printSiteMeta bool
var printCrossRefMeta bool

var DoiCmd = &cobra.Command{
	Use:   "doi [doi1 doi2 doi3...]",
	Short: "Can be used to access files via DOI.",
	Long:  `Can be used to access files via DOI. More see here https://github.com/openanno/bget.`,
	Run: func(cmd *cobra.Command, args []string) {
		doiCmdRunOptions(cmd, args)
	},
}

func doiCmdRunOptions(cmd *cobra.Command, args []string) {
	initCmd(cmd, args)
	checkArgs(cmd, "doi")
	checkDownloadDir(bgetClis.Doi != "")
	if bgetClis.Doi != "" || bgetClis.ListFile != "" {
		downloadDoi(cmd, args)
		bgetClis.HelpFlags = false
	}
	if bgetClis.HelpFlags {
		cmd.Help()
	}
}

func downloadDoi(cmd *cobra.Command, args []string) {
	var sem = make(chan bool, bgetClis.Thread)
	var urls = []string{}
	var destDirArray []string
	var doi = parseArgsDoi()
	for _, v := range doi {
		sem <- true
		go func(v string) {
			defer func() {
				<-sem
			}()
			urlsTmp, opt := doiSpiders(v)
			*urlsTmp = slice.DropSliceDup(*urlsTmp)
			var outDir = path.Join(bgetClis.DownloadDir, v)
			if opt != nil && opt.PrintSiteMeta {
				outputSiteMetaData(outDir, opt)
			}
			if opt != nil && opt.PrintCrossRefMeta {
				outputCrossRefData(outDir, opt, cmd, args)
			}
			for range *urlsTmp {
				destDirArray = append(destDirArray, outDir)
			}
			urls = append(urls, *urlsTmp...)
		}(v)
	}
	for i := 0; i < cap(sem); i++ {
		sem <- true
	}
	netOpt := setNetParams(&bgetClis)
	cnet.HTTPGetURLs(urls, destDirArray, netOpt)
}

func parseArgsDoi() (doi []string) {
	if bgetClis.Doi != "" && strings.Contains(bgetClis.Doi, bgetClis.Seperator) {
		doi = strings.Split(bgetClis.Doi, bgetClis.Seperator)
	} else if bgetClis.Doi != "" && strings.Contains(bgetClis.Doi, "\n") {
		doi = strings.Split(bgetClis.Doi, "\n")
	} else if bgetClis.Doi != "" {
		doi = []string{bgetClis.Doi}
	} else if bgetClis.ListFile != "" {
		doi = cio.ReadLines(bgetClis.ListFile)
	}
	return doi
}

func outputSiteMetaData(outDir string, opt *spider.DoiSpiderOpt) {
	var err error
	var of io.Writer
	var buf *bytes.Buffer
	metaJSON, _ := json.Marshal(*opt.CitationMeta)
	log.Infof("Website metadata of %s: %s", opt.Doi, string(metaJSON))
	outfn := path.Join(outDir, "website.meta.json")
	log.Infof("Saving website metadata of %s => %s", opt.Doi, outfn)
	cio.CreateFileParDir(outfn)
	if cio.CheckFilesOverwrite([]string{outfn}, bgetClis.Overwrite) == -1 {
		of, err = cio.Open(outfn)
		if err != nil {
			log.Warnln(err)
		} else {
			indent := ""
			for i := 0; i < 2; i++ {
				indent = indent + " "
			}
			opt := pretty.Options{
				Indent:   indent,
				SortKeys: false,
			}
			buf = bytes.NewBuffer(pretty.PrettyOptions(metaJSON, &opt))
			io.Copy(of, buf)
		}
	}
}

func outputCrossRefData(outDir string, opt *spider.DoiSpiderOpt,
	cmd *cobra.Command, args []string) {
	var err error
	var crossRefEndp = types.CrossRefEndpoints{}
	crossRefEndp.Doi.Doi = opt.Doi
	var bapiClis = types.BapiClisT{
		TaskID:       bgetClis.TaskID,
		Verbose:      bgetClis.Verbose,
		SaveLog:      bgetClis.SaveLog,
		LogDir:       bgetClis.LogDir,
		HelpFlags:    false,
		Proxy:        bgetClis.Proxy,
		Timeout:      bgetClis.Timeout,
		RetSleepTime: bgetClis.RetSleepTime,
		Retries:      bgetClis.Retries,
		PrettyJSON:   false,
		XML2json:     true,
	}
	var buf bytes.Buffer
	var of = io.Writer(&buf)
	var outfn string
	fetch.CrossRef(&crossRefEndp, &bapiClis, func() { initCmd(cmd, args) }, of)
	log.Infof("Crossref metadata of %s: %s", opt.Doi, string(buf.Bytes()))
	outfn = path.Join(outDir, "crossref.citation.json")
	log.Infof("Saving crossref metadata of %s => %s", opt.Doi, outfn)
	cio.CreateFileParDir(outfn)
	if cio.CheckFilesOverwrite([]string{outfn}, bgetClis.Overwrite) == -1 {
		of, err = cio.Open(outfn)
		if err != nil {
			log.Warnln(err)
		} else {
			indent := ""
			for i := 0; i < 2; i++ {
				indent = indent + " "
			}
			opt := pretty.Options{
				Indent:   indent,
				SortKeys: false,
			}
			bufTmp := bytes.NewBuffer(pretty.PrettyOptions(buf.Bytes(), &opt))
			io.Copy(of, bufTmp)
		}
	}
}

func doiSpiders(doi string) (urls *[]string, opt *spider.DoiSpiderOpt) {
	urls = &[]string{}
	if !strings.Contains(doi, "/") {
		return urls, opt
	}
	if stringo.StrDetect(doi, "http[s]://doi.org/") {
		doi = stringo.StrReplaceAll(doi, "http[s]://doi.org/", "")
	}
	doiTmp := strings.Split(doi, "/")
	doiOrg := doiTmp[0]
	var t int
	var citationMeta = make(map[string]string)
	opt = &spider.DoiSpiderOpt{
		Doi:               doi,
		Proxy:             bgetClis.Proxy,
		Timeout:           bgetClis.Timeout,
		FullText:          fullText == "true",
		PrintSiteMeta:     printSiteMeta,
		PrintCrossRefMeta: printCrossRefMeta,
		CitationMeta:      &citationMeta,
		Supplementary:     suppl,
	}
	if pmc {
		*urls = spider.PmcSpider(opt)
		if len(*urls) == 0 && !suppl {
			return urls, opt
		}
	}
	if len(*urls) == 0 && scihub {
		*urls = spider.ScihupSpider(opt)
		if len(*urls) > 0 && !suppl {
			return urls, opt
		}
	}
	if len(*urls) > 0 {
		opt.FullText = false
	}
	client := cnet.NewHTTPClient(opt.Timeout, opt.Proxy)
	req, _ := http.NewRequest("HEAD", "https://doi.org/"+opt.Doi, nil)
	resp, err := client.Do(req)
	if err != nil && strings.Contains(err.Error(), "http") {
		link := stringo.StrExtract(err.Error(), `".*"`, 1)[0]
		link = stringo.StrReplaceAll(link, `"`, "")
		u, _ := neturl.Parse(link)
		opt.URL = u
	} else {
		defer resp.Body.Close()
		u, _ := neturl.Parse(resp.Request.URL.String())
		opt.URL = u
	}
	for k := range spider.DoiSpidersPool {
		if k == doiOrg {
			for t = 0; t < bgetClis.Retries+1; t++ {
				*urls = append(*urls, spider.DoiSpidersPool[doiOrg](opt)...)
				if len(*urls) == 0 {
					if bgetClis.Retries != 0 {
						log.Warnf("Pool spider: %s returns empty, on attempt %d... retrying after %d seconds.", doi, t+1, bgetClis.RetSleepTime)
						time.Sleep(time.Duration(bgetClis.RetSleepTime) * time.Second)
					} else {
						log.Warnf("Pool spider: %s returns empty, on attempt %d...", doi, t+1)
					}
					continue
				} else {
					break
				}
			}
		}
	}
	if len(*urls) == 0 && universeSpider {
		for t = 0; t < bgetClis.Retries; t++ {
			*urls = append(*urls, spider.UniVersalDoiSpider(opt)...)
			if len(*urls) == 0 {
				if bgetClis.Retries != 0 {
					log.Warnf("Universal spider: %s returns empty, on attempt %d... retrying after %d seconds.", doi, t+1, bgetClis.RetSleepTime)
					time.Sleep(time.Duration(bgetClis.RetSleepTime) * time.Second)
				} else {
					log.Warnf("Pool spider: %s returns empty, on attempt %d...", doi, t+1)
				}
				continue
			} else {
				break
			}
		}
	}
	return urls, opt
}

func init() {
	universeSpider = true
	DoiCmd.Flags().BoolVarP(&scihub, "enable-scihub", "", false, "enable try scihub spider.")
	DoiCmd.Flags().BoolVarP(&pmc, "enable-pmc", "", false, "enable try PMC database.")
	DoiCmd.Flags().StringVarP(&fullText, "full-text", "", "true", "access full text.")
	DoiCmd.Flags().BoolVarP(&suppl, "suppl", "", false, "access supplementary files.")
	DoiCmd.Flags().BoolVarP(&printSiteMeta, "print-meta", "", false, "print website meta data.")
	DoiCmd.Flags().BoolVarP(&printCrossRefMeta, "print-crossref", "", false, "print crossref meta data.")

	setGlobalFlag(DoiCmd, &bgetClis)
	setKeyListFlag(DoiCmd, &bgetClis, "dois")
	exampleXML2Json := "`bget api ncbi --xml2json --json-pretty -q '30487223[pmid] or 30402350[pmid] or 29279377[pmid]' --size 3 -m 3 | grep / | grep 10. | sed 's/ .* \"//' | tr -d '\",' | sort -u`"
	DoiCmd.Example = fmt.Sprintf(`  bget doi 10.5281/zenodo.3363060 10.5281/zenodo.3357455 10.5281/zenodo.3351812 -t 3
  bget doi 10.1016/j.devcel.2017.03.001 10.1016/j.stem.2019.07.009 10.1016/j.celrep.2018.03.072 -t 2
	
  # query pdf and meta data using PubMed ID
  dois=%s && echo ${dois} && bget doi ${dois} --print-meta --print-crossref
  
  bget doi 10.1080/15548627.2018.1505155 --proxy http://username:password@hostname:port
  bget doi 10.1182/blood.2019000200 --enable-scihub
  bget doi 10.1109/JPROC.2019.2905423 -n
  bget doi 10.1073/pnas.1814397115 --print-meta --print-crossref`, exampleXML2Json)
}
